module/zcommon/zfs_fletcher.c

```diff
@@ -245,7 +214,107 @@ fletcher_4_incremental_byteswap(const void *buf, uint64_t size,
 	ZIO_SET_CHECKSUM(zcp, a, b, c, d);
 }
 
+static const struct fletcher_4_calls fletcher_4_generic_calls = {
+	.init = fletcher_4_generic_init,
+	.compute = fletcher_4_generic,
+	.compute_byteswap = fletcher_4_generic_byteswap,
+	.name = "generic"
+};
+
+static const struct fletcher_4_calls *chosen = &fletcher_4_generic_calls;
+
+void
+fletcher_4_native(const void *buf, uint64_t size, zio_cksum_t *zcp)
+{
+	chosen->init(zcp);
+	chosen->compute(buf, size, zcp);
+	if (chosen->fini != NULL)
+		chosen->fini(zcp);
+}
+
+void
+fletcher_4_byteswap(const void *buf, uint64_t size, zio_cksum_t *zcp)
+{
+	chosen->init(zcp);
+	chosen->compute_byteswap(buf, size, zcp);
+	if (chosen->fini != NULL)
+		chosen->fini(zcp);
+}
+
+void
+fletcher_4_incremental_native(const void *buf, uint64_t size,
+    zio_cksum_t *zcp)
+{
+	fletcher_4_generic(buf, size, zcp);
+}
+
+void
+fletcher_4_incremental_byteswap(const void *buf, uint64_t size,
+    zio_cksum_t *zcp)
+{
+	fletcher_4_generic_byteswap(buf, size, zcp);
+}
+
+
 #if defined(_KERNEL) && defined(HAVE_SPL)
+
+extern struct fletcher_4_calls fletcher_4_avx2_calls;
+
+static const struct fletcher_4_calls *fletcher_4_algos[] = {
+	&fletcher_4_generic_calls,
+#if defined(HAVE_AVX) && defined(HAVE_AVX2)
+	&fletcher_4_avx2_calls,
+#endif
+};
```

Comment:
To prevent leaking arch specific stuff into common code.
We can do something like this.
```
fletcher_4_algos[] = {
        &fletcher_4_generic_calls,
#include <asm/fletcher.h>
};
```
fletcher.h can in turn include fletcher_xxx.h according to the arch.
And put `fletcher_4_avx2_calls,` inside fletcher_x86.h

Just in case if there will be other arch adding their SIMD stuff.